### PR TITLE
 [JENKINS-19887] - Allow time zone to be set on a per user basis

### PR DESCRIPTION
**Pull request by nagrubb**
See [JENKINS-19887](https://issues.jenkins-ci.org/browse/JENKINS-19887).

### Proposed changelog entries

* Entry 1: Add TimeZone as configurable on per-user basis

### CODE REVIEW

File: core/src/main/java/hudson/Functions.java
Comment (jvz) [inline core/src/main/java/hudson/Functions.java]: Please don't do this.

File: core/src/main/resources/hudson/model/TimeZoneProperty/config.jelly
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/model/TimeZoneProperty/config.jelly]: Please adjust the affiliation here or just put "Jenkins contributors"

File: core/src/main/resources/hudson/model/TimeZoneProperty/config.properties
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/model/TimeZoneProperty/config.properties]: same

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/TimeZoneProperty.java]:
```suggestion
    @Symbol("timezone")
```

No need in uppercase Z IMHO

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/TimeZoneProperty.java]: Would be nice to add some Javadoc to the new class

File: core/src/main/java/hudson/Functions.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Functions.java]: I suggest annotating methods properly so that nullability is explicit

File: core/src/main/java/hudson/Functions.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Functions.java]: What will happen here if the string cannot be converted? Looks like some error handling and protective code is missing

**Review by oleg-nenashev — DISMISSED** [review]
Thanks @silent-snowman ! It would be really great to finally get this feature over the line. It is 2019, but definitely better late then never.

I am requesting changes, because the code  does not look to be robust against invalid timezone IDs. They can be easily passed by Jenkins configuration-as-code plugin with this change, and from what I see the code may start failing with runtime exceptions after it.  So some checks should be added. Bonus points for proper annotations and Javadoc

File: core/src/main/java/hudson/Functions.java
Comment (silent-snowman) [reply core/src/main/java/hudson/Functions.java]: Sorry Intellij does this automatically, but will revert it.

File: core/src/main/java/hudson/Functions.java
Comment (silent-snowman) [reply core/src/main/java/hudson/Functions.java]: Per Java docs, it'll just use GMT.

Comment (silent-snowman) [reply core/src/main/java/hudson/Functions.java]: (Also just tested to confirm)

**Comment by silent-snowman** [comment]
@oleg-nenashev  I understand your initial response as that was mine when initially looking at java.util.TimeZone API since it's most common for APIs in Java to throw exceptions on invalid/unexpected input, but that's not the case for TimeZone.getTimeZone() at least. This method will not throw an exception but instead just return the default timezone of GMT/UTC. I think this is pretty reasonable, but if you don't think so than please let me know how you'd prefer to handle it. After your comment I double checked by just testing out putting an invalid input directly in a sample user/config.xml and it behaved exactly as the Java Docs define to just return GMT and no exceptions will be thrown thus there isn't really much error handling needed as I think this is pretty reasonable. Thoughts?
https://docs.oracle.com/javase/7/docs/api/java/util/TimeZone.html#getTimeZone(java.lang.String)

File: core/src/main/java/hudson/Functions.java
Comment (silent-snowman) [reply core/src/main/java/hudson/Functions.java]: added

**Comment by oleg-nenashev** [comment]
> java.util.TimeZone API since it's most common for APIs in Java to throw exceptions on invalid/unexpected input, but that's not the case for TimeZone.getTimeZone() at least

Right, I should have read Javadoc. It is so unusual for Oracle APIs that I did not even check such possibility before commenting :(  Comment retracted.

File: core/src/main/java/hudson/Functions.java
Comment (varyvol) [reply core/src/main/java/hudson/Functions.java]: Shouldn't it use the system one, which is the one this is overriding?

**Review by alecharp — CHANGES_REQUESTED** [review]
It would be great if you could add some tests and shows (with before/after screenshots) the impact of this PR. Also, it seems the default timezone used is the system one, but you also mentioned the timezone fallback on GMT.

File: core/src/main/java/hudson/Functions.java
Comment (silent-snowman) [reply core/src/main/java/hudson/Functions.java]: @varyvol actually that could be debated. In terms of Jenkins, using the system one if I have an invalid user preference vs using GMT makes zero difference to me personally. The company I'm at now has offices in PST, MST, EET, ICT. It doesn't make my life any easier if it's one of those I'm not in vs GMT. I think I could do it, but not sure I understand the benefit for introducing this extra bit of complexity. 

@alecharp I'll see about adding a unit test. I have to figure out the Jenkins testing framework.

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (silent-snowman) [reply core/src/main/java/hudson/model/TimeZoneProperty.java]: I actually had it as Default, but that is actually more confusing in my opinion so I changed it to blank which also makes the code cleaner. The way it's blank now actually makes more sense to me.

**Comment by silent-snowman** [comment]
I'll take some screen shots and all them to this PR.

**Comment by silent-snowman** [comment]
Before change (or without logging in)
Screen Shot 2019-07-17 at 9 49 14 PM

After change (or logging in with user setting)
Screen Shot 2019-07-17 at 9 51 25 PM

User configuration page:
Screen Shot 2019-07-17 at 9 52 21 PM

**Comment by silent-snowman** [comment]
There are other pages affected too, I don't think this captures everything but it's a start and this would be very benefit to my current workplace that's across 5+ very different timezones.

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (silent-snowman) [reply core/src/main/java/hudson/model/TimeZoneProperty.java]: added

**Comment by silent-snowman** [comment]
Can someone help me on how I can trigger a specific test? Right now I'm doing everything all at once and that's way too slow (i.e. using `maven install` to do everything is painful). I don't work with maven, Java or even web development professionally so honestly even getting this far was painful.

File: core/src/main/java/hudson/Functions.java
Comment (varyvol) [reply core/src/main/java/hudson/Functions.java]: @silent-snowman maybe the benefits are not great (I guess it's debatable depending on specific situations) but I think changing the current behaviour would actually cause strangeness at least.

**Comment by varyvol** [comment]
@silent-snowman you should be able to do something like "mvn test -Dtest=<testClass>#<testMethod>"

**Comment by silent-snowman** [comment]
@varyvol @alecharp 

In all honesty I think it's quite silly but I added tests to ensure the Java framework doesn't change the behavior of the TimeZone API to throw an exception in the future. If Oracle did this, they should deprecate the old API and write a new one because that's the only responsible way to extend a very fundamental API. The tests for this particular feature do look pretty though, but whether they will catch any sort of regression in the future, I highly doubt.

I don't agree that reverting to the timezone that the Jenkins server is configured to is the right way to handle an invalid time zone for three reasons:
1) This introduces quite a bit complexity because there are two ways to set the TimeZone of Jenkins:
a) Via "Jelly config" using `-Dorg.apache.commons.jelly.tags.fmt.timeZone=<timezone>`
b) By setting the TimeZone of the server. 
If someone chooses to do a) then TimeZone.getDefault() will still return the time zone of the server it's running on and not the time zone specified for jelly.
2) Being a user of this feature to me GMT is actually less arbitrary than whatever it is that the maintainer configured the Jenkins server's timezone to. In all honesty if you have a team across multiple timezones (which is what this feature is targeted at) and you have Jenkins set to anything but GMT, you are kind of a shitty person. Also I've made it so that when using this feature, the time zone is appended in the build history so it's quite obvious what your timezone is set to to verify the feature is working correctly. Not to mention it's also in the footer of Jenkins as it uses timeFormat=long (along with anywhere else that uses this). It's actually quite bad practice to not include the TimeZone in every printed time unless it's GMT...
3) It's more standard even within Jenkins to revert to GMT on an unrecognized timezone. For example, if I launch Jenkins with `-Dorg.apache.commons.jelly.tags.fmt.timeZone=InvalidTimeZone`, it will revert to GMT rather than the TimeZone of the machine/server it's running on.

**Comment by silent-snowman** [comment]
Rather than introduce this complexity that is to handle an edge case error, instead I added a warning log statement in case for some reason doesn't immediately notice the wrong timezone through the UI they should notice through the logger.

**Comment by silent-snowman** [comment]
@varyvol I know what you meant and I fundamentally disagree that adding complexity for this is a good idea especially because it's to handle an edge case. Also I don't want to invest anymore time in this to figure out how to get the timezone from Jelly. I don't get paid to work on Jenkins so I'm not going to do this change because I honestly don't see the benefit or agree. I added a log statement so if this is happening it's quite clear. I don't think it's confusing at all to revert to GMT especially given that timezones are displayed.

**Comment by silent-snowman** [comment]
@varyvol let me show you how shitty what you are proposing looks like and then you know if you are happy then go ahead and merge it.

File: core/src/main/java/hudson/Functions.java
Comment (silent-snowman) [reply core/src/main/java/hudson/Functions.java]: @varyvol I think this change is terrible, but whatever it's not my project. I hate that you are making me put my name next to a shitty change because it seems like you won't approve the PR without it. Not to mention it's quite nitpicky.

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (silent-snowman) [reply core/src/main/java/hudson/model/TimeZoneProperty.java]: I don't think this change, but I made it, so please approve the PR.

**Comment by silent-snowman** [comment]
@varyvol @jvz @alecharp @daniel-beck @oleg-nenashev could you please approve this? I made all the changes per suggestions even though I don't agree with them, but ultimately it's not my project and I really hate that Jenkins doesn't have decent TimeZone support.

File: core/src/main/java/hudson/Functions.java
Comment (jvz) [inline core/src/main/java/hudson/Functions.java]: You know how you could avoid dealing with invalid values? Provide only valid values that can be used. Use [`ZoneId.getAvailableZoneIds()`](https://docs.oracle.com/javase/8/docs/api/java/time/ZoneId.html#getAvailableZoneIds--) to get a list of all the timezones that are possible currently. Then you can use an `<f:select/>` jelly component with a `doFillTimeZoneChoiceValues()` type function.

By using only valid timezones, now you won't fall back to anything. Let me know if you'd like a more concrete example or a PR to your PR or something.

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (jvz) [inline core/src/main/java/hudson/model/TimeZoneProperty.java]: So I have a few alternative ways you could keep the time zone valid without having to use that ugly workaround.

1. Modify this constructor to check if the timezone is valid; if not, throw an `IllegalArgumentException`.
2. Not sure if this works, but change the type of the field, constructor argument, and getter, to `TimeZone`, and then add a `writeReplace()`/`ProxyObject::readResolve()` pair to serialize it since `java.util.TimeZone` is not whitelisted for serialization.
3. Alternatively, maybe we can add `java.util.TimeZone` to the whitelisted classes for serialization combined with the first part of 2.

Comment (silent-snowman) [reply core/src/main/java/hudson/model/TimeZoneProperty.java]: It's not about keeping it valid, it's already not possible to be invalid but this is error handling if for some reason it was changed through configuration as code. This checks if it's valid and is the cleanest and most efficient way I think to check if it's valid.

Comment (jvz) [reply core/src/main/java/hudson/model/TimeZoneProperty.java]: Oh, I see. Then I think adding the validation in the constructor here would make more sense than trying to recover from a failure that shouldn't happen in practice (except for typos in hand-written config as code, but that's always a given with yaml files).

Comment (silent-snowman) [reply core/src/main/java/hudson/model/TimeZoneProperty.java]: Also your ideas run counter to your coworkers suggestion. I give up, this is all honestly quite frustrating and I don't want to do this in my free time. I'm going to investigate using TeamCity or just living with Jenkins as is.

Comment (jvz) [reply core/src/main/java/hudson/model/TimeZoneProperty.java]: Coworkers disagree; what else is new? This same debate would be had if we all worked at different companies. ;)

File: core/src/main/resources/hudson/model/TimeZoneProperty/config.jelly
Comment (jvz) [inline core/src/main/resources/hudson/model/TimeZoneProperty/config.jelly]:
```suggestion
  <f:entry field="timeZoneName" title="${%Time Zone}">
```

This will make the string localizable.

**Review by jvz — APPROVED** [review]
Looks good to me, though I haven't tested this manually. Thanks for contributing!

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (silent-snowman) [reply core/src/main/java/hudson/model/TimeZoneProperty.java]: Fair, but just difficult being on the outside as to what the actual guideline is for handling errors in Jenkins, but I'm guessing like most open source there's no particular standard.

File: core/src/main/java/hudson/Functions.java
Comment (daniel-beck) [inline core/src/main/java/hudson/Functions.java]: Wouldn't something like `isUserTimeZoneOverride` be more in line with the usual naming convention?

Comment (daniel-beck) [inline core/src/main/java/hudson/Functions.java]: Additions to the API either need Javadoc and `@since TODO` (to be filled in later), or `@Restricted(NoExternalUse.class)`

File: core/src/main/java/hudson/Functions.java
Comment (daniel-beck) [inline core/src/main/java/hudson/Functions.java]: Same

File: core/src/main/java/hudson/Functions.java
Comment (daniel-beck) [inline core/src/main/java/hudson/Functions.java]: Same

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/TimeZoneProperty.java]: Remove

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/TimeZoneProperty.java]: `@since TODO` or `@Restricted`

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/TimeZoneProperty.java]: Unsure whether the naming of `TimeZoneProperty#forCurrentUser()` isn't too clever. `getTimeZoneForCurrentUser` and done?

**Review by daniel-beck — COMMENTED** [review]
Overall I like it.

I wonder however whether it wouldn't be easier to introduce something like this perhaps in a plugin. Except the newly added time zone suffix in the widget (which is pretty much a hack to begin with) it should be possible to inject the time zone into `l:layout` via extension point.

Have alternative UIs been considered, such as Blue Ocean, or additional occurrences of date/time information contributed e.g. via Pipeline plugins?
Was a systematic review of views at least in core done to determine whether this sufficiently covers all built-in UI?

**Comment by oleg-nenashev** [comment]
> I wonder however whether it wouldn't be easier to introduce something like this perhaps in a plugin. Except the newly added time zone suffix in the widget (which is pretty much a hack to begin with) it should be possible to inject the time zone into l:layout via extension point.

Benefit from that is that a plugin could target older Jenkins core versions so that plugins rendering times do not need a Jenkins core bump. On the other hand, I am not sure how complex it would be to introduce a new extension point. 

Needs discussion, CC @jenkinsci/core to get more opinions. Personally I do not mind putting it to the core. It is a common use-case (one may say that having user email is also a common use-case, but it it in the mailer plugin).

> Have alternative UIs been considered, such as Blue Ocean, or additional occurrences of date/time information contributed e.g. via Pipeline plugins?

CC @halkeye who might clarify whether we could pick up the change with the current REST APIs

**Comment by fcojfernandez** [comment]
> Needs discussion, CC @jenkinsci/core to get more opinions. Personally I do not mind putting it to the core. It is a common use-case (one may say that having user email is also a common use-case, but it it in the mailer plugin).

Jenkins should manage the timezone differences, IMHO. The underlying situation is more and more common and making the admin users to install a plugin should be avoided.

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (silent-snowman) [reply core/src/main/java/hudson/model/TimeZoneProperty.java]: This follows the convention from another user property (I forget which). Being this is a public static it should be called TimeZoneProperty.forCurrentUser() which I think is quite explanatory.

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (silent-snowman) [reply core/src/main/java/hudson/model/TimeZoneProperty.java]: What exactly do you want removed? @daniel-beck

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (silent-snowman) [reply core/src/main/java/hudson/model/TimeZoneProperty.java]: Could you clarify? Do you just wanted me to add @Restricted to the class?

**Comment by oleg-nenashev** [comment]
I will see whether I can get it over the line next week. Sorry, DW-JW and infra issues

**Comment by rsandell** [comment]
@oleg-nenashev seems this ball is in your court?

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (batmat) [reply core/src/main/java/hudson/model/TimeZoneProperty.java]: Remove the `@author` please. We do not generally use them anywhere. This is generally misleading, and git blame is better if we need to ask questions.

**Review by rsandell — DISMISSED** [review]
Review comments from @daniel-beck are still not addressed.

New public method lacking javadoc, `@since  ` and possible rename.

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (daniel-beck) [reply core/src/main/java/hudson/model/TimeZoneProperty.java]: Either add `@since` to indicate when this was added to the public API of Jenkins, or `@Restricted` to prevent use in plugins. No strong opinion which.

**Comment by daniel-beck** [comment]
> Jenkins should manage the timezone differences, IMHO. The underlying situation is more and more common and making the admin users to install a plugin should be avoided.

A plugin may actually facilitate adoption, as plugin maintainers aren't always quick to require recent core baselines.

**Comment by silent-snowman** [comment]
@daniel-beck addressed all of your comments

**Comment by silent-snowman** [comment]
@batmat @varyvol @rsandell any love? I really would like to see this in Jenkins.

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (varyvol) [inline core/src/main/java/hudson/model/TimeZoneProperty.java]: Is this line doing something?

Comment (silent-snowman) [reply core/src/main/java/hudson/model/TimeZoneProperty.java]: I don't think so, must be an accident.

**Review by varyvol — APPROVED** [review]

File: core/src/main/java/hudson/Functions.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Functions.java]: We may need to unrestrict it later, fine for now

File: core/src/main/java/hudson/Functions.java
Comment (oleg-nenashev) [reply core/src/main/java/hudson/Functions.java]:
```suggestion
    @CheckForNull
```

**Review by oleg-nenashev — APPROVED** [review]
Just minor comments, it is ready to go IMHO

**Comment by daniel-beck** [comment]
@silent-snowman Thanks for addressing my specific code review findings and your patience with review feedback.

I cannot find a response either in a comment, or in code, to my top-level comments, quoted below:

> Have alternative UIs been considered, such as Blue Ocean, or additional occurrences of date/time information contributed e.g. via Pipeline plugins?

The API additions are all `@Restricted`. How would plugins not using Jelly to render their UI make use of this, notably Blue Ocean? Installed on 20% of instances, it's notable enough by now that we should offer it a path to adopt this.

> Was a systematic review of views at least in core done to determine whether this sufficiently covers all built-in UI?

FWIW I count 11 uses of `l:formatDate` in core alone and only one gets the suffix. Why?

TBH I'm not a huge fan of this suffix, it looks like it would just introduce consistency problems, as there are dozens of uses of this control across plugins.

I was unable to find in the discussion whether it was your idea or added in response to a review comment. I apologize if you added it in response to a review comment; sometimes feedback is contradictory especially if there's no clear best solution.

I'm sorry that this proposal seems stuck in review hell. This PR changes a pretty basic part of the Jenkins UI, and that's always one of the most difficult kind of change, as we need to consider the larger plugin ecosystem around Jenkins, and how it would adapt to a given change.

**Comment by silent-snowman** [comment]
@daniel-beck I don't know why I keep looking at this PR because it just upsets me and makes me not want to use Jenkins even though I honestly quite like it. I spent my own time to do this change because I really wanted it and really thought it'd make Jenkins work better for global teams. I get zero money for this and this was purely done on my own time. I feel like I'm getting negative respect for this contribution as well as no help from people that actually are paid to work on Jenkins.

Right now, no. I'm not an active contributor towards Jenkins and honestly I learned how to make this change because I really wanted it. I don't think it makes sense to hold up a PR that does move the needle forward on Jenkins UI if only just a little bit. Sure it doesn't expose this functionality to plugins, but it also doesn't regress their functionality. 

This doesn't cover all of the uses of formatDate, I can do other PRs to help fix that. However, I'm a bigger fan of doing small changes since they are easier to review. This PR is evidence of this taking over 3 months to get reviews for. 

Can we really just get this in and I can help move this forward? I just Jenkins quite a bit at work and this would really help me.

**Comment by michaelneale** [comment]
Whilst blue ocean is 20% stats wise, if this doesn't break things, and the worst case is some slight inconsistency, I would still love to see this in. As someone who works in weird timezones where most of the day it isn't even the same day in everywhere else, I applaud timezone friendly improvements. 

If this goes in, inconsistencies can be addressed right? this doesn't break things, and materially makes things better for many users all over the world - so why not?

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (res0nance) [inline core/src/main/java/hudson/model/TimeZoneProperty.java]:
```suggestion
    @CheckForNull
```

**Review by res0nance — APPROVED** [review]
Just a minor suggestion. Overall I like the feature and really want it in Jenkins. I have users across more than 5 timezones and this will certainly help. My master is not even in my own timezone and I have had to deal with computing the right time in my head all the time.

**Review by timja — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
I plan to merge it tomorrow if no negative feedback

File: core/src/main/java/hudson/model/TimeZoneProperty.java
Comment (oleg-nenashev) [reply core/src/main/java/hudson/model/TimeZoneProperty.java]: @res0nance I would like to land it. Since it is a minot change, let's leave it for a follow-up PR by somebody

**Comment by oleg-nenashev** [comment]
Thanks a lot for your contribution and patience @silent-snowman ! It should land in the next weekly

**Comment by oleg-nenashev** [comment]
It might have been merged prematurely. I may revert it after talking with stakeholders in PM.
@jenkinsci/core @jenkinsci/core-pr-reviewers If anyone has reservations here, just comment and I will put it back to review

**Comment by jsoref** [comment]
I'm curious about the details.

The only thing that bothers me about this was the UI for configuring it. I started working on a different UI but got distracted. It doesn't help that there really are no good UIs for this problem space. (macOS and friends tend to use a map, and at least macOS has some really bizarre quirks wrt its map.)

**Comment by daniel-beck** [comment]
Caused [JENKINS-61972](https://issues.jenkins-ci.org/browse/JENKINS-61972).

**Comment by silent-snowman** [comment]
I'll take a look.

**Pull request merged**
